platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,927 @@
+- Accept a list of paths to not display `SiPIndicator`
+- Provide a React Context with the following
+  - `saveForm` function
+    - Save the form when `saveForm` is called
+  - `loadForm` function
+    - Call the supplied endpoint with the supplied arguments
+    - Set the form data to the form data in the result
+    - Perform any data migrations applicable
+    - Navigate to the page supplied by the result
+- Accept a [`dataMigrations`](#data-migrations) prop to be used with `loadForm`
+
+##### Component: `SiPIndicator`
+**TODO:** Rename this thing. Ideas?
+
+**Purpose:** To indicate when the form has been auto-saved and to provide a link
+to save and exit the form.
+
+**Usage requirements:**
+This must appear as a descendant of `SaveInProgress` since it uses the React
+Context provided by that component to save the form.
+
+**Responsibilities:**
+- Display a link to save and exit the form
+- Call an API endpoint to save the form data
+- Navigate to `/form-saved`
+
+**Location on the page:** This component will appear at the bottom of every page
+except for those it is configured not to. The default routes which it will not
+appear will be:
+- `/introduction`
+- `/review-and-submit`
+- `/confirmation`
+- `/form-saved`
+
+**A note on usage:** The primary usage of this component is internal to
+Formulate, used by `SaveInProgress`, but it may be used separately if desired.
+
+##### Component: `FormSaved`
+**Purpose:** To make the common `/form-saved` route easy to add to an
+application.
+
+**Usage requirements:**
+_None_
+
+**Responsibilities:**
+- Render some text
+- Render a link to continue the form
+
+##### Data migrations
+When a form in production needs to be changed in a way that will affect the
+user's ability to complete the form accurately, a data migration may be written
+to account for it. Examples include:
+- The validation for a field changes
+- An existing field was previously not required but now is
+- A new required field is added
+- A new page is added
+- A field is removed
+
+The goal of migrations is to bring the saved data of an old version of the form
+up to match the new version of the form.
+
+A `dataMigrations` prop may be passed to the `SaveInProgress` component.
+`dataMigrations` must be an array of functions. Each function will be called
+with the form data and metadata, either from the API request or a previous
+migration function. Metadata will include:
+- Form version when the data was saved
+- The page URL when the form was saved
+
+Each function must return the new form data and metadata. If multiple migration
+functions must be run, the result of an earlier migration will be passed to the
+next migration in the array.
+
+Migrations are run during the `loadForm` function. It will check the form
+version in the metadata against the length of the `dataMigrations` array. If the
+latter is greater, all migrations after the saved form version will be called in
+order.
```

Comment:
For form 526, the list of rated disabilities is reloaded upon each form start, and if the loaded list doesn't match the SiP list, we redirect them back to the rated disabilities page so that they know a change happened and that they need to review their options. So, maybe this `dataMigrations` can allow a way to reroute, or change the `page URL` metadata

Reply:
Oooh, this is interesting. That isn't a data migration since the form itself didn't change. The data from an external source did. It sounds like you need the ability to provide your own `loadForm` function (using the one provided by Formulate, likely). One option is to consume the form state React Context and override the `loadForm` function, but that seems cumbersome. I'll see if I can't figure out a cleaner way for you to supply your own `loadForm`. :+1: 

This is a great call out, thanks! :raised_hands:

Reply:
After thinking about it some more, I think the best developer experience will be to provide a `postLoad` function that occurs between data migration and routing to the `metadata.returnUrl`. This can be an `async` function in which you can make an external service call (fetching the latest rated disabilities, for example) and modify either the form data or metadata (re-routing back to the rated disabilities selection page, for example).